feat(new): Added Azure.AKS.MaintenanceWindow (#2997)
Azure/PSRule.Rules.Azure: docs/en/rules/Azure.AKS.MaintenanceWindow.md

## RECOMMENDATION

Consider using the customer-controlled maintenance windows 'aksManagedAutoUpgradeSchedule' for all cluster upgrade scenarios and 'aksManagedNodeOSUpgradeSchedule' for all node OS security patching scenarios.

## EXAMPLES

### Configure with Azure template

To deploy AKS clusters that pass this rule:

- Deploy two maintenance configurations.
 - Set the `name` to `aksManagedAutoUpgradeSchedule` and `aksManagedNodeOSUpgradeSchedule`.

For example:

```json
{
  "type": "Microsoft.ContainerService/managedClusters",
  "apiVersion": "2024-03-02-preview",
  "name": "[parameters('clusterName')]",
  "location": "[parameters('location')]",
  "sku": {
    "name": "Basic",
    "tier": "Standard"
  },
  "identity": {
    "type": "SystemAssigned"
  },
  "properties": {
    "autoUpgradeProfile": {
      "upgradeChannel": "stable",
      "nodeOSUpgradeChannel": "NodeImage"
    },
    "dnsPrefix": "[parameters('dnsPrefix')]",
    "agentPoolProfiles": [
      {
        "name": "agentpool",
        "osDiskSizeGB": "[parameters('osDiskSizeGB')]",
        "count": "[parameters('agentCount')]",
        "vmSize": "[parameters('agentVMSize')]",

>         "osType": "Linux",
>         "mode": "System"
>       }
>     ],
>     "linuxProfile": {
>       "adminUsername": "[parameters('linuxAdminUsername')]",
>       "ssh": {
>         "publicKeys": [
>           {
>             "keyData": "[parameters('sshRSAPublicKey')]"
>           }
>         ]
>       }
>     }
>   }
> },
> {
>   "type": "Microsoft.ContainerService/managedClusters/maintenanceConfigurations",
>   "apiVersion": "2024-03-02-preview",
>   "name": "[format('{0}/{1}', parameters('clusterName'), 'aksManagedAutoUpgradeSchedule')]",
>   "properties": {
>     "maintenanceWindow": {
>       "schedule": {
>         "weekly": {
>           "intervalWeeks": 1,
>           "dayOfWeek": "Sunday"
>         }
>       },
>       "durationHours": 4,
>       "utcOffset": "+00:00",
>       "startDate": "2024-07-15",
>       "startTime": "00:00"
>     }
>   },
>   "dependsOn": [
>     "[resourceId('Microsoft.ContainerService/managedClusters', parameters('clusterName'))]"
>   ]
> },
> {
>   "type": "Microsoft.ContainerService/managedClusters/maintenanceConfigurations",
>   "apiVersion": "2024-03-02-preview",
>   "name": "[format('{0}/{1}', parameters('clusterName'), 'aksManagedNodeOSUpgradeSchedule')]",
>   "properties": {
>     "maintenanceWindow": {
>       "schedule": {
>         "weekly": {
>           "intervalWeeks": 1,
>           "dayOfWeek": "Sunday"
>         }
>       },
>       "durationHours": 4,
>       "utcOffset": "+00:00",
>       "startDate": "2024-07-15",
>       "startTime": "00:00"
>     }
>   },
>   "dependsOn": [
>     "[resourceId('Microsoft.ContainerService/managedClusters', parameters('clusterName'))]"
>   ]
> }

Reviewer: ```suggestion
{
  "type": "Microsoft.ContainerService/managedClusters/maintenanceConfigurations",
  "apiVersion": "2024-03-02-preview",
  "name": "[format('{0}/{1}', parameters('clusterName'), 'aksManagedAutoUpgradeSchedule')]",
  "properties": {
    "maintenanceWindow": {
      "schedule": {
        "weekly": {
          "intervalWeeks": 1,
          "dayOfWeek": "Sunday"
        }
      },
      "durationHours": 4,
      "utcOffset": "+00:00",
      "startDate": "2024-07-15",
      "startTime": "00:00"
    }
  },
  "dependsOn": [
    "[resourceId('Microsoft.ContainerService/managedClusters', parameters('clusterName'))]"
  ]
}
```

```

> 

Reviewer: ```suggestion
For node OS auto-upgrades see the example:

```json
{
  "type": "Microsoft.ContainerService/managedClusters/maintenanceConfigurations",
  "apiVersion": "2024-03-02-preview",
  "name": "[format('{0}/{1}', parameters('clusterName'), 'aksManagedNodeOSUpgradeSchedule')]",
  "properties": {
    "maintenanceWindow": {
      "schedule": {
        "weekly": {
          "intervalWeeks": 1,
          "dayOfWeek": "Sunday"
        }
      },
      "durationHours": 4,
      "utcOffset": "+00:00",
      "startDate": "2024-07-15",
      "startTime": "00:00"
    }
  },
  "dependsOn": [
    "[resourceId('Microsoft.ContainerService/managedClusters', parameters('clusterName'))]"
  ]
}
```

```

### Configure with Bicep

To deploy AKS clusters that pass this rule:

> - Deploy two maintenance configurations.
>  - Set the `name` to `aksManagedAutoUpgradeSchedule` and `aksManagedNodeOSUpgradeSchedule`.
> 
> For example:

Reviewer: ```suggestion
- Deploy maintenance configurations for cluster version and node OS auto-upgrades.
  - For cluster version auto-upgrades, set the `name` property to `aksManagedAutoUpgradeSchedule`.
  - For node OS auto-upgrades, set the `name` property to `aksManagedNodeOSUpgradeSchedule`.

For cluster version auto-upgrades see the example:
```

```bicep

> resource aks 'Microsoft.ContainerService/managedClusters@2024-03-02-preview' = {
>   name: clusterName
>   location: location
>   sku: {
>     name: 'Basic'
>     tier: 'Standard'
>   }
>   identity: {
>     type: 'SystemAssigned'
>   }
>   properties: {
>     autoUpgradeProfile: {
>       upgradeChannel: 'stable'
>       nodeOSUpgradeChannel: 'NodeImage'
>     }
>     dnsPrefix: dnsPrefix
>     agentPoolProfiles: [
>       {
>         name: 'agentpool'
>         osDiskSizeGB: osDiskSizeGB
>         count: agentCount
>         vmSize: agentVMSize
>         osType: 'Linux'
>         mode: 'System'
>       }
>     ]
>     linuxProfile: {
>       adminUsername: linuxAdminUsername
>       ssh: {
>         publicKeys: [
>           {
>             keyData: sshRSAPublicKey
>           }
>         ]
>       }
>     }
>   }
> }

Reviewer: ```suggestion
```

resource aksManagedAutoUpgradeSchedule 'Microsoft.ContainerService/managedClusters/maintenanceConfigurations@2024-03-02-preview' = {
  parent: aks
  name: 'aksManagedAutoUpgradeSchedule'
  properties: {
    maintenanceWindow: {
      schedule: {
        weekly: {
          intervalWeeks: 1
          dayOfWeek: 'Sunday'
        }
      }
      durationHours: 4
      utcOffset: '+00:00'
      startDate: '2024-07-15'
      startTime: '00:00'
    }
  }

> }
> 
> resource aksManagedNodeOSUpgradeSchedule 'Microsoft.ContainerService/managedClusters/maintenanceConfigurations@2024-03-02-preview' = {
>   parent: aks
>   name: 'aksManagedNodeOSUpgradeSchedule'
>   properties: {
>     maintenanceWindow: {
>       schedule: {
>         weekly: {
>           intervalWeeks: 1
>           dayOfWeek: 'Sunday'
>         }
>       }
>       durationHours: 4
>       utcOffset: '+00:00'
>       startDate: '2024-07-15'
>       startTime: '00:00'
>     }
>   }
> }

Reviewer: ```suggestion
}
```

For node OS auto-upgrades see the example:

```bicep
resource aksManagedNodeOSUpgradeSchedule 'Microsoft.ContainerService/managedClusters/maintenanceConfigurations@2024-03-02-preview' = {
  parent: aks
  name: 'aksManagedNodeOSUpgradeSchedule'
  properties: {
    maintenanceWindow: {
      schedule: {
        weekly: {
          intervalWeeks: 1
          dayOfWeek: 'Sunday'
        }
      }
      durationHours: 4
      utcOffset: '+00:00'
      startDate: '2024-07-15'
      startTime: '00:00'
    }
  }
}
```
